File: android/StelliveHubAndroid/app/src/main/java/dev/minepacu/stelliveeventnotifier/feature/hubevents/HubEventLinkPolicy.kt
package dev.minepacu.stelliveeventnotifier.feature.hubevents

import dev.minepacu.stelliveeventnotifier.core.model.HubEvent
import dev.minepacu.stelliveeventnotifier.core.model.HubEventLink
import dev.minepacu.stelliveeventnotifier.core.model.HubEventLinkKind
import dev.minepacu.stelliveeventnotifier.core.model.HubEventScheduleItem
import java.net.URI

enum class HubEventLinkCtaMode { NONE, DIRECT, SHEET }

object HubEventLinkPolicy {
    fun resolvedEventLinks(event: HubEvent): List<HubEventLink> =
        resolve(event.links).ifEmpty {
            resolve(listOfNotNull(
                legacy("legacy:${event.id}:purchase", HubEventLinkKind.PURCHASE, event.purchaseUrl, null, 0),
                legacy("legacy:${event.id}:ticket", HubEventLinkKind.TICKET, event.ticketUrl, null, 1),
                legacy("legacy:${event.id}:source", HubEventLinkKind.SOURCE, event.sourceUrl, event.sourceLabel, 2),
            ))
        }

    fun resolvedScheduleLinks(item: HubEventScheduleItem): List<HubEventLink> =
        resolve(item.links).ifEmpty {
            val actionKind = when (item.kind) {
                dev.minepacu.stelliveeventnotifier.core.model.HubEventScheduleKind.SALES_OPEN -> HubEventLinkKind.PURCHASE
                dev.minepacu.stelliveeventnotifier.core.model.HubEventScheduleKind.TICKET_OPEN -> HubEventLinkKind.TICKET
                dev.minepacu.stelliveeventnotifier.core.model.HubEventScheduleKind.DEADLINE,
                dev.minepacu.stelliveeventnotifier.core.model.HubEventScheduleKind.MAIN_WINDOW -> HubEventLinkKind.RESERVATION
                dev.minepacu.stelliveeventnotifier.core.model.HubEventScheduleKind.ANNOUNCEMENT -> HubEventLinkKind.SOURCE
                else -> HubEventLinkKind.CONTENT
            }
            resolve(listOfNotNull(
                legacy("legacy:${item.id}:action", actionKind, item.actionUrl, null, 0),
                legacy("legacy:${item.id}:source", HubEventLinkKind.SOURCE, item.sourceUrl, item.sourceLabel, 1),
            ))
        }

    fun displayLinkLabel(link: HubEventLink): String =
        link.label?.trim()?.takeIf(String::isNotEmpty) ?: displayKindLabel(link.kind)

    fun displayKindLabel(kind: HubEventLinkKind): String = when (kind) {
        HubEventLinkKind.SOURCE -> "출처"
        HubEventLinkKind.PURCHASE -> "구매"
        HubEventLinkKind.TICKET -> "티켓"
        HubEventLinkKind.RESERVATION -> "예약"
        HubEventLinkKind.CONTENT -> "콘텐츠"
        HubEventLinkKind.VIDEO -> "영상"
        HubEventLinkKind.MAP -> "지도"
        HubEventLinkKind.CUSTOM -> "관련 링크"
    }

    fun eventCtaMode(event: HubEvent): HubEventLinkCtaMode = when (resolvedEventLinks(event).size) {
        0 -> HubEventLinkCtaMode.NONE
        1 -> HubEventLinkCtaMode.DIRECT
        else -> HubEventLinkCtaMode.SHEET
    }

    fun resolvedExpandedScheduleItemIds(
        previousEventId: String?,
        eventId: String,
        currentIds: Set<String>,
        highlightedScheduleItemId: String?,
    ): Set<String> = buildSet {
        if (previousEventId == eventId) addAll(currentIds)
        highlightedScheduleItemId?.let(::add)
    }

    fun effectivePrimaryScheduleItemId(event: HubEvent): String? = event.scheduleItems
        .asSequence()
        .filter { it.cancelledAt == null && it.isPrimary }
        .sortedWith(
            compareBy<HubEventScheduleItem> { it.sortOrder }
                .thenBy { it.startsAt }
                .thenBy { it.createdAt }
                .thenBy { it.id }
        )
        .firstOrNull()
        ?.id

    private fun resolve(links: List<HubEventLink>): List<HubEventLink> {
        val seen = mutableSetOf<String>()
        return links
            .mapNotNull { link -> link.copy(url = link.url.trim(), label = link.label?.trim()?.takeIf(String::isNotEmpty)).takeIf { isHttps(it.url) } }
            .sortedWith(compareBy<HubEventLink> { it.sortOrder }.thenBy { it.id })
            .filter { seen.add(it.url) }
    }

    private fun legacy(id: String, kind: HubEventLinkKind, url: String?, label: String?, sortOrder: Int): HubEventLink? =
        url?.trim()?.takeIf(::isHttps)?.let { HubEventLink(id, kind, label?.trim()?.takeIf(String::isNotEmpty), it, sortOrder) }

    private fun isHttps(value: String): Boolean = runCatching {
        val uri = URI(value)
        uri.scheme.equals("https", ignoreCase = true) && !uri.host.isNullOrBlank()
    }.getOrDefault(false)
}

File: backend/stellive-hub-api/src/hub-events/hubEventLinkPolicy.ts
import type { HubEvent, HubEventLink, HubEventLinkKind, HubEventScheduleItem } from "../types.js";

export const hubEventLinkKinds: readonly HubEventLinkKind[] = [
  "source",
  "purchase",
  "ticket",
  "reservation",
  "content",
  "video",
  "map",
  "custom"
];

export type HubEventLinkInput = Omit<HubEventLink, "id" | "createdAt" | "updatedAt"> &
  Partial<Pick<HubEventLink, "id" | "createdAt" | "updatedAt">>;

function isHttpsUrl(value: string): boolean {
  try {
    return new URL(value).protocol === "https:";
  } catch {
    return false;
  }
}

export function normalizeHubEventLinks(links: readonly HubEventLinkInput[] | undefined): HubEventLinkInput[] | undefined {
  if (links === undefined) return undefined;
  return links.map((link, index) => ({
    ...link,
    label: link.label?.trim() || undefined,
    url: link.url?.trim(),
    sortOrder: link.sortOrder ?? index
  }));
}

function resolvedLinks(links: readonly HubEventLink[] | undefined): HubEventLink[] {
  const seen = new Set<string>();
  return (links ?? [])
    .filter((link) => isHttpsUrl(link.url.trim()))
    .map((link) => ({ ...link, label: link.label?.trim() || undefined, url: link.url.trim() }))
    .sort((left, right) => left.sortOrder - right.sortOrder || (left.id ?? "").localeCompare(right.id ?? ""))
    .filter((link) => {
      if (seen.has(link.url)) return false;
      seen.add(link.url);
      return true;
    });
}

function legacyLink(id: string, kind: HubEventLinkKind, url: string | undefined, label: string | undefined, sortOrder: number): HubEventLink | undefined {
  const normalizedUrl = url?.trim();
  if (!normalizedUrl || !isHttpsUrl(normalizedUrl)) return undefined;
  return { id, kind, label: label?.trim() || undefined, url: normalizedUrl, sortOrder };
}

export function resolvedEventLinks(event: Pick<HubEvent, "id" | "links" | "purchaseUrl" | "ticketUrl" | "sourceUrl" | "sourceLabel">): HubEventLink[] {
  const current = resolvedLinks(event.links);
  if (current.length > 0) return current;
  return resolvedLinks([
    legacyLink(`legacy:${event.id}:purchase`, "purchase", event.purchaseUrl, undefined, 0),
    legacyLink(`legacy:${event.id}:ticket`, "ticket", event.ticketUrl, undefined, 1),
    legacyLink(`legacy:${event.id}:source`, "source", event.sourceUrl, event.sourceLabel, 2)
  ].filter((link): link is HubEventLink => Boolean(link)));
}

export function resolvedScheduleLinks(item: Pick<HubEventScheduleItem, "id" | "kind" | "links" | "actionUrl" | "sourceUrl" | "sourceLabel">): HubEventLink[] {
  const current = resolvedLinks(item.links);
  if (current.length > 0) return current;
  const actionKind: HubEventLinkKind = item.kind === "sales_open"
    ? "purchase"
    : item.kind === "ticket_open"
      ? "ticket"
      : item.kind === "deadline" || item.kind === "main_window"
        ? "reservation"
        : item.kind === "announcement"
          ? "source"
          : "content";
  return resolvedLinks([
    legacyLink(`legacy:${item.id}:action`, actionKind, item.actionUrl, undefined, 0),
    legacyLink(`legacy:${item.id}:source`, "source", item.sourceUrl, item.sourceLabel, 1)
  ].filter((link): link is HubEventLink => Boolean(link)));
}

export function firstEventLegacyProjection(links: readonly HubEventLinkInput[] | undefined): Pick<HubEvent, "purchaseUrl" | "ticketUrl"> {
  const sorted = (links ?? []).slice().sort((left, right) => (left.sortOrder ?? 0) - (right.sortOrder ?? 0) || (left.id ?? "").localeCompare(right.id ?? ""));
  return {
    purchaseUrl: sorted.find((link) => link.kind === "purchase" || link.kind === "reservation")?.url,
    ticketUrl: sorted.find((link) => link.kind === "ticket")?.url
  };
}

export function firstScheduleLegacyProjection(links: readonly HubEventLinkInput[] | undefined): Pick<HubEventScheduleItem, "actionUrl" | "sourceUrl" | "sourceLabel"> {
  const sorted = (links ?? []).slice().sort((left, right) => (left.sortOrder ?? 0) - (right.sortOrder ?? 0) || (left.id ?? "").localeCompare(right.id ?? ""));
  const source = sorted.find((link) => link.kind === "source");
  const action = sorted.find((link) => link.kind !== "source");
  return {
    actionUrl: action?.url,
    sourceUrl: source?.url,
    sourceLabel: source?.label
  };
}
